File: go/backoff/backoff.go
package backoff

import (
	"math"
	"math/rand"
)

// Compute returns exponential backoff with jitter.
func Compute(attempt int, initialSeconds float64, maxSeconds float64, jitter float64) float64 {
	if attempt < 1 {
		attempt = 1
	}
	exp := math.Min(maxSeconds, initialSeconds*math.Pow(2, float64(attempt-1)))
	if jitter <= 0 {
		return exp
	}
	delta := exp * jitter
	return math.Max(0, exp+(rand.Float64()*2*delta-delta))
}

File: python/workflow/backoff.py
"""
Exponential backoff with jitter.
"""

from __future__ import annotations

import random


def compute_backoff_seconds(
    attempt: int,
    initial_seconds: float,
    max_seconds: float,
    jitter: float,
) -> float:
    """
    Compute exponential backoff with jitter in seconds.

    attempt is 1-based (first retry attempt).
    """
    exp = min(max_seconds, initial_seconds * (2 ** max(0, attempt - 1)))
    if jitter <= 0:
        return exp
    # Apply jitter as +/- percentage.
    delta = exp * jitter
    return max(0.0, exp + random.uniform(-delta, delta))
